File: 2019/day4/part2.py
#!/usr/bin/env python3

def has_double(num):
    prev = -1
    count = 1
    for i in num:
        if int(i) == prev:
            count += 1
        else:
            if count == 2:
                return True
            count = 1
        prev = int(i)

    if count == 2:
        return True
    return False


def decreases(num):
    prev = -1
    for i in num:
        if int(i) < prev:
            return True
        prev = int(i)
    return False


with open('input', 'r') as f:
    lower, upper = map(int, f.read().strip().split('-'))

count = 0
for i in range(upper - lower):
    candidate = str(lower + i)
    if has_double(candidate) and not decreases(candidate):
        print(f'{candidate} is okay')
        count += 1
    else:
        if not decreases(candidate):
            print(f'{candidate} does not decrease but is not okay')

print(f'The answer is: {count}')

File: 2019/day4/part2.go
package main

import (
	"fmt"
	"strconv"
	"strings"
)

func has_double(s string) bool {
	prev := -1
	count := 0
	okay := false

	for _, r := range s {
		n, _ := strconv.Atoi(string(r))
		if n == prev {
			count++
			if count == 2 {
				okay = true
			} else {
				okay = false
			}
		} else {
			if count == 2 {
				return true
			}
			okay = false
			count = 1
		}

		prev = n
	}

	return okay
}

func decreases(s string) bool {
	prev := -1
	for _, r := range s {
		n, _ := strconv.Atoi(string(r))
		if n < prev {
			return true
		}
		prev = n
	}
	return false
}

func main() {
	input := "265275-781584"
	meh := strings.Split(input, "-")

	lower, _ := strconv.Atoi(meh[0])
	upper, _ := strconv.Atoi(meh[1])

	count := 0
	for i := lower + 1; i < upper; i++ {
		num := strconv.Itoa(i)
		if has_double(num) && !decreases(num) {
			count++
		}
	}

	fmt.Printf("The answer is %d\n", count)
}
